include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -283,14 +286,17 @@ def AndOfPad : Pat <
   [(KnownWidth $x), (UIntType $x), (EqualTypes $x, $pad)]>;
 
 /// cvt is adding a 0, which will pass through the and
-/// and(cvt(x:U), const) -> cat(0, and(x,trunc(const))
+/// Only trigger when cvt is full length
+/// and(cvt(x:U), const) -> cat(0, and(x,sextortrunc(const))
 def AndCvtU : Pat <
-  (AndPrimOp:$old (CvtPrimOp $x), (ConstantOp:$ocst $cst)),
+  (AndPrimOp:$old (CvtPrimOp:$cvt $x), (ConstantOp:$ocst $cst)),
   (MoveNameHint $old, (CatPrimOp 
     (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), APSInt(APInt()))"> $ocst),
-    (AndPrimOp $x, (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), cast<IntType>($0.getType()), $1.getValue().trunc($1.getValue().getBitWidth()-1).zext(*cast<IntType>($0.getType()).getWidth()))"> $x, $cst))
+    (AndPrimOp $x, (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), "
```

Comment:
Classic "which location does this get" dilemma.  One alternative is to give it `$1.getLoc()` but as I think we've discussed before arguments both ways.

Reply:
I think constants should never have locations when they are synthesized.  But that is a thing to audit for at some point.